tests/test_server/test_lineage/test_job_lineage.py

```diff
@@ -258,38 +315,105 @@ async def test_get_job_lineage(
                 "end_reason": run.end_reason,
             }
             for run in sorted(runs, key=lambda x: x.id)
+        ],
+    }
+
+
+async def test_get_job_lineage_with_direction_and_until(
+    test_client: AsyncClient,
+    async_session: AsyncSession,
+    lineage: LINEAGE_FIXTURE_ANNOTATION,
+):
+    all_jobs, all_runs, all_operations, all_datasets, all_inputs, _ = lineage
+
+    job = all_jobs[0]
+
+    since = min(run.created_at for run in all_runs if run.job_id == job.id)
+    until = since + timedelta(seconds=1)
+
+    raw_runs = [run for run in all_runs if run.job_id == job.id and since <= run.created_at <= until]
+    run_ids = {run.id for run in raw_runs}
+    assert raw_runs
+
+    inputs = [input for input in all_inputs if input.run_id in run_ids and since <= input.created_at <= until]
+    assert inputs
+
+    # Only operations with some inputs are returned
+    operation_ids = {input.operation_id for input in inputs}
+    operations = [operation for operation in all_operations if operation.id in operation_ids]
+    assert operations
+
+    # Same for runs
+    run_ids = {operation.run_id for operation in operations}
+    runs = [run for run in all_runs if run.id in run_ids]
+    assert runs
```

Comment:
```suggestion
    # Only runs with some inputs are 
    run_ids = {input.run_id for input in inputs}
    runs = [run for run in all_runs if run.id in run_ids]
    assert runs
```
all_operations can be dropped